Objects/regions.c

```diff
@@ -643,9 +675,294 @@ PyObject* _Py_MakeImmutable(PyObject* obj)
+    }
+}
+
+// Open method (sets the region to "open")
+static PyObject *PyRegion_open(PyRegionObject *self, PyObject *args) {
+    regionmetadata_open(self->metadata);
+    Py_RETURN_NONE;  // Return None (standard for methods with no return value)
+}
+
+// Close method (sets the region to "closed")
+static PyObject *PyRegion_close(PyRegionObject *self, PyObject *args) {
+    regionmetadata_close(self->metadata);  // Mark as closed
+    Py_RETURN_NONE;  // Return None (standard for methods with no return value)
+}
+
+// Adds args object to self region
+static PyObject *PyRegion_add_object(PyRegionObject *self, PyObject *args) {
+    if (!args) {
+        Py_RETURN_NONE;
+    }
+
+    regionmetadata* md = PyRegion_get_metadata(self);
+    if (is_default_region(Py_GET_REGION(args))) {
+        Py_SET_REGION(args, (Py_uintptr_t) md);
+        Py_RETURN_NONE;
+    } else {
+        PyErr_SetString(PyExc_RuntimeError, "Object already had an owner or was immutable!");
+        return NULL;
+    }
+}
+
+// Remove args object to self region
+static PyObject *PyRegion_remove_object(PyRegionObject *self, PyObject *args) {
+    if (!args) {
+        Py_RETURN_NONE;
+    }
+
+    regionmetadata* md = PyRegion_get_metadata(self);
+    if (Py_GET_REGION(args) == (Py_uintptr_t) md) {
+        Py_SET_REGION(args, _Py_DEFAULT_REGION);
+        Py_RETURN_NONE;
+    } else {
+        PyErr_SetString(PyExc_RuntimeError, "Object not a member of region!");
+        return NULL;
+    }
+}
+
+// Return True if args object is member of self region
+static PyObject *PyRegion_owns_object(PyRegionObject *self, PyObject *args) {
+    if ((Py_uintptr_t) PyRegion_get_metadata(self) == Py_GET_REGION(args)) {
+        Py_RETURN_TRUE;
+    } else {
+        Py_RETURN_FALSE;
+    }
+}
+
+static PyObject *PyRegion_repr(PyRegionObject *self) {
+    regionmetadata* data = self->metadata;
+    // FIXME: deprecated flag, but config.parse_debug seems to not work?
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+    if (Py_DebugFlag) {
+#pragma GCC diagnostic pop
+        // Debug mode: include detailed representation
+        return PyUnicode_FromFormat(
+            "Region(lrc=%d, osc=%d, name=%S, is_open=%d)", data->lrc, data->osc, self->name ? self->name : Py_None, data->is_open
+        );
+    } else {
+        // Normal mode: simple representation
+        return PyUnicode_FromFormat("Region(name=%S, is_open=%d)", self->name ? self->name : Py_None, data->is_open);
+    }
+    Py_RETURN_NONE;
+}
+
+// Define the RegionType with methods
+static PyMethodDef PyRegion_methods[] = {
+    {"open", (PyCFunction)PyRegion_open, METH_NOARGS, "Open the region."},
+    {"close", (PyCFunction)PyRegion_close, METH_NOARGS, "Close the region."},
+    {"is_open", (PyCFunction)PyRegion_is_open, METH_NOARGS, "Check if the region is open."},
+    {"add_object", (PyCFunction)PyRegion_add_object, METH_O, "Add object to the region."},
```

Comment:
I don't think you need `_object` in this methods. Also, in the past we used `setattr` and `getattr` to add things to the region and then get them back out again. How would that work? It doesn't seem possible to query a region about its contents at the moment, either as a sequence or a mapping.

Reply:
This is just meant as a temporary debug function to test the invariant etc. I think it makes sense to have this as a simple method and wait until the write barrier works. I'll add a comment saying that they're temporary